subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1282 @@
+				? USB_SPEED_FS : USB_SPEED_HS;
+
+		data->if_state = IF_STATE_SPEED_SENT;
+
+		ncm_notify_speed_change.header.wIndex = sys_cpu_to_le16(cdc_ncm_get_ctrl_if(data));
+		ncm_notify_speed_change.downlink = sys_cpu_to_le32(usb_speed);
+		ncm_notify_speed_change.uplink   = sys_cpu_to_le32(usb_speed);
+		ret = _usbd_cdc_ncm_send_notification(dev, &ncm_notify_speed_change,
+				sizeof(ncm_notify_speed_change));
+		LOG_DBG("cdc_ncm_send_notification_speed_change %d", ret);
+	} else if (data->if_state == IF_STATE_SPEED_SENT) {
+		data->if_state = IF_STATE_DONE;
+
+		ncm_notify_connected.header.wIndex = sys_cpu_to_le16(cdc_ncm_get_ctrl_if(data));
+		ret = _usbd_cdc_ncm_send_notification(dev, &ncm_notify_connected,
+				sizeof(ncm_notify_connected));
+		LOG_DBG("cdc_ncm_send_notification_connected %d", ret);
+	}
+}   /* _usbd_cdc_ncm_notification_next_step */
+
+
+
+static int usbd_cdc_ncm_request(struct usbd_class_data *const c_data,
+		struct net_buf *buf, int err)
+/**
+ * Endpoint request completion event handler: handle NCM request from host.
+ */
+{
+	struct usbd_contex *uds_ctx = usbd_class_get_ctx(c_data);
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct udc_buf_info *bi;
+
+	bi = udc_get_buf_info(buf);
+	LOG_DBG("ep: 0x%02x", bi->ep);
+
+	if (bi->ep == cdc_ncm_get_bulk_out(c_data)) {
+		/* data received */
+		return cdc_ncm_acl_out_cb(c_data, buf, err);
+	}
+
+	if (bi->ep == cdc_ncm_get_bulk_in(c_data)) {
+		LOG_DBG("free sync_sem");
+		k_sem_give(&data->sync_sem);
+		return 0;
+	}
+
+	if (bi->ep == cdc_ncm_get_int_in(c_data)) {
+		LOG_DBG("notification");
+		_usbd_cdc_ncm_notification_next_step(c_data);
+		return 0;
+	}
+
+	return usbd_ep_buf_free(uds_ctx, buf);
+}   /* usbd_cdc_ncm_request */
+
+
+
+static void usbd_cdc_ncm_update(struct usbd_class_data *const c_data,
+		const uint8_t iface, const uint8_t alternate)
+/**
+ * Configuration update handler.
+ * \param alternate   == 0 -> NCM reset (spec 7.2), == 1 -> normal data exchange (spec 5.3).
+ * According to spec 7.1, first \a ConnectionSpeedChange and the \a NetworkConnection
+ * have to be sent.
+ */
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct usbd_cdc_ncm_desc *desc = data->desc;
+	const uint8_t first_iface = desc->if0.bInterfaceNumber;
+
+	LOG_DBG("New configuration, interface %u alternate %u first %u",
+			iface, alternate, first_iface);
+
+	if (iface == first_iface + 1) {
+		LOG_DBG("set alt: %d", alternate);
+		data->itf_data_alt = alternate;
+	}
+
+	if (iface != first_iface + 1  ||  alternate == 0) {
+		LOG_DBG("Skip iface %u alternate %u", iface, alternate);
+
+		/*
+		 * reset internal status
+		 */
+		data->tx_sequence = 0;
+		data->if_state = IF_STATE_FIRST_SKIPPED;
+		return;
+	}
+
+	if (data->if_state == IF_STATE_INIT) {
+		data->if_state = IF_STATE_FIRST_SKIPPED;
+		LOG_DBG("Skip first iface enable");
```

Comment:
Any help for understanding this? Is there a bug elsewhere that needs a workaround, or is it handling of a special case?

Reply:
Yes... and no... explanation can be found in the NCM spec, chapter 7.1.  I have to admit that this does not seem to be intuitive, but this is how NCM works

Reply:
> 7.1 Notification Sequencing 
> NCM functions are required to send ConnectionSpeedChange and NetworkConnection notifications in a specific order. To simplify the coding of host device drivers, functions that are going to send a NetworkConnection notification with wValue == 0001h must first send a ConnectionSpeedChange notification that  indicates the connection speed that will be in effect when the new connection takes effect. This sequencing is justified as follows.  If the ConnectionSpeedChange follows a NetworkConnection notification, then the host driver cannot signal network connection with the correct speed until the ConnectionSpeedChange is received.  This delay may introduce latency between bus events and system events, or may cause host system overhead due to a spurious change in speed.  If the function signals the connection speed first, then the host driver will know the signaling speed at the time the network connection becomes valid.

Reply:
It looks like it skips through the states of a state machine without looking at the content to confirm this was the right state. Is it due to this driver controlling what events are generated here?
https://github.com/zephyrproject-rtos/zephyr/pull/72310/files#diff-d6af60862ee4693cfe35b5df0806df2fdfff1921acc16a65b23ee407548e7212R546-R572

Many thanks!

Reply:
The host driver sends notifications in a "special" sequence with alternate=0/1.  The sequence is (more or less) from the spec.  If the device driver is in the correct state, it has to respond first with "speed change" and then with "connected" notifications.
After that the driver is allowed to transmit bulk data.

Reply:
For some reason, I thought `IF_STATE_FIRST_SKIPPED` was to skip the first message: drop it without processing it, but no: it is skipping any action on the first *alternate setting*! Right?

If interested in being more dummy-proof, renaming `IF_STATE_FIRST_SKIPPED` to `IF_STATE_FIRST_ALTERNATE_SETTING_SKIPPED` or shorter that still hint about the `_FIRST_...` of what it is.

> ### 5.3 Data Interface Descriptor Requirements 
> The Data Interface of an NCM networking function shall have two alternate settings.  **The first alternate  setting** (the default interface setting, alternate setting 0) shall include no endpoints and therefore **no networking traffic can be exchanged when the default interface setting is selected**.  The second alternate setting (alternate setting 1) is used for normal operation, and shall include one bulk IN endpoint and one bulk OUT endpoint.

Reply:
done (have to push)